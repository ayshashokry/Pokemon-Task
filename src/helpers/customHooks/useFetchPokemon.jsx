import axios from "axios";
import React, { useEffect, useState } from "react";

export function useFetchPokemon(url, initialState, isDetails) {
  const [data, setData] = useState(initialState);
  const [pokeSpices, setPokeSpices] = useState({});

  const [isLoading, setLoading] = useState(false);
  const [error, setError] = useState(null);

  const fetchData = async () => {
    setLoading(true);

    await axios
      .get(url)
      .then((response) => {
        setData(response?.data);
        if (isDetails) {
          axios
            .get(response?.data?.species?.url)
            .then((result) => setPokeSpices(result.data));
        }
        setLoading(false);
        setError(null);
      })
      .catch((error) => {
        setError(error);
        setLoading(false);
      });
  };
  useEffect(() => {
    let cleanup = false;
    fetchData();
    return () => {
      cleanup = true;
    };
  }, [url]);
  return [data, isLoading, error, pokeSpices];
}
