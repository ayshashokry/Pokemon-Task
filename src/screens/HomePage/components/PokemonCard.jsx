import { Button } from "antd";
import React from "react";
import { Link } from "react-router-dom";
import { useFetchPokemon } from "../../../helpers/customHooks/useFetchPokemon";
import { Card } from "antd";import { LazyLoadImage } from "react-lazy-load-image-component";

export default function PokemonCard({ pokemon }) {
  const url = pokemon.url;
  const [pokemonDetails, , , pokeSpices] = useFetchPokemon(url, {}, true);
  const pokemonName =
    pokemon?.name?.charAt(0)?.toUpperCase() + pokemon?.name?.slice(1);
  const renderLink = () => (
    <Link to={`/pokemon/${pokemonDetails?.id}`}>
      <img
        loading="lazy"
        src={pokemonDetails?.sprites?.front_default}
        alt="pokemonImage"
      />
    </Link>
  );
  return (
    <Card
      className="mx-xl-4  mx-lg-2 mx-md-3 pokemonCard mb-4"
      cover={renderLink()}>
      <h4>{pokemonName}</h4>
      

      {pokeSpices?.flavor_text_entries !== undefined && (
        <p>{pokeSpices?.flavor_text_entries[0]?.flavor_text}</p>
      )}
      <div className="pokeTypes">
        {pokemonDetails.types?.length > 0 &&
          pokemonDetails.types?.map((type) => (
            <Button
              style={{ backgroundColor: pokeSpices?.color?.name }}
              key={type?.slot}>
              {type?.type?.name?.charAt(0)?.toUpperCase() +
                type?.type?.name?.slice(1)}
            </Button>
          ))}
      </div>
    </Card>
  );
}
