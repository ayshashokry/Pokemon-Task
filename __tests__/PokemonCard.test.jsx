import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { vi } from "vitest";

import * as exports from "../src/helpers/customHooks/useFetchPokemon";
import PokemonCard from "../src/screens/HomePage/components/PokemonCard";
const mocked_data = {
  items: [
    {
      id: 1,
      name: "name",
      types: [
        {
          slot: 1,
          type: { name: "grass", url: "mocked url" },
        },
        {
          slot: 2,
          type: { name: "poison", url: "mocked url" },
        },
      ],
      sprites: { front_default: "mock data" },
      url: "mocked url",
    },
    ,
    ,
    { flavor_text_entries: [{ flavor_text: "mock data" }] },
  ],
};

describe("Pokemon Card", () => {
  test("it should render correctly", () => {
    vi.spyOn(exports, "useFetchPokemon").mockReturnValue(mocked_data?.items);
    const element = render(
      <MemoryRouter>
        <PokemonCard pokemon={mocked_data?.items[0]} />
      </MemoryRouter>
    );
    expect(element).toMatchSnapshot();
  });
});
