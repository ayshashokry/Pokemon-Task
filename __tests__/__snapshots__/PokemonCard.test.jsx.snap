// Vitest Snapshot v1

exports[`Pokemon Card > it should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ant-card ant-card-bordered mx-xl-4  mx-lg-2 mx-md-3 pokemonCard mb-4 css-dev-only-do-not-override-ixblex"
      >
        <div
          class="ant-card-cover"
        >
          <a
            href="/pokemon/1"
          >
            <img
              alt="pokemonImage"
              src="mock data"
            />
          </a>
        </div>
        <div
          class="ant-card-body"
        >
          <h4>
            Name
          </h4>
          <p>
            mock data
          </p>
          <div
            class="pokeTypes"
          >
            <button
              class="ant-btn css-dev-only-do-not-override-ixblex ant-btn-default"
              type="button"
            >
              <span>
                Grass
              </span>
            </button>
            <button
              class="ant-btn css-dev-only-do-not-override-ixblex ant-btn-default"
              type="button"
            >
              <span>
                Poison
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ant-card ant-card-bordered mx-xl-4  mx-lg-2 mx-md-3 pokemonCard mb-4 css-dev-only-do-not-override-ixblex"
    >
      <div
        class="ant-card-cover"
      >
        <a
          href="/pokemon/1"
        >
          <img
            alt="pokemonImage"
            src="mock data"
          />
        </a>
      </div>
      <div
        class="ant-card-body"
      >
        <h4>
          Name
        </h4>
        <p>
          mock data
        </p>
        <div
          class="pokeTypes"
        >
          <button
            class="ant-btn css-dev-only-do-not-override-ixblex ant-btn-default"
            type="button"
          >
            <span>
              Grass
            </span>
          </button>
          <button
            class="ant-btn css-dev-only-do-not-override-ixblex ant-btn-default"
            type="button"
          >
            <span>
              Poison
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
